stdlib/strtod.c

```diff
@@ -15,184 +15,539 @@
+static const char *floatparse_zerox(const char *str, bool *is_hex)
+{
+	*is_hex = 0;
+	if (*str == '0') {
 		str++;
+		if (((*str == 'x') || (*str == 'X')) && valid_hex(str + 1)) {
+			str++;
+			*is_hex = 1;
+		}
+
+		while (*str == '0') {
+			str++;
+		}
+
+		return str;
+	}
+	else if (valid_decimal(str)) {
+		return str;
 	}
 
-	/*  check the sign */
-	if (*str == '-' && (isdigit(*(str + 1)) || *(str + 1) == '.')) {
-		negative = 1;
-		str++;
+	return NULL;
+}
+
+static int get_digit_value(char byte, bool hex)
+{
+	int c = byte;
+	c -= 'A';
+	if (hex && (c >= 0)) {
+		c &= ~0x20;
+		c += 10;
+		if (c > 0xf) {
+			return -1;
+		}
 	}
-	else if (*str == '+' && (isdigit(*(str + 1)) || *(str + 1) == '.')) {
-		str++;
+	else {
+		c = byte - '0';
+		if (c > 9) {
+			return -1;
+		}
 	}
 
-	/*  get the integer part */
-	while (isdigit(*str)) {
-		if (len_int < numdig) {
-			whole_number = whole_number * base + (*str - '0');
+	return c;
+}
+
+
+static const char *scan_number(const char *str, bool hex, bool allow_dec, uint64_t limit, size_t *n_decimal, size_t *n_overflowed, uint64_t *out)
+{
+	const int base = hex ? 16 : 10;
+	uint64_t result = 0u;
+	bool got_dec = !allow_dec;
+	while (1) {
+		char c = *str;
+		if (!got_dec && (c == DECIMAL_POINT)) {
+			got_dec = true;
+			str++;
+			continue;
 		}
-		else {
-			overflow_int++;
+
+		int dig = get_digit_value(c, hex);
+		if (dig < 0) {
+			break;
 		}
 
-		if (whole_number) {
-			len_int++;
+		if (result <= limit) {
+			result = result * base + dig;
+			if (got_dec) {
+				(*n_decimal)++;
+			}
+		}
+		else if (allow_dec != got_dec) {
+			(*n_overflowed)++;
 		}
 
-		is_number = 1;
 		str++;
 	}
 
-	/*  get the fractiontion part */
-	if (*str == '.') {
-		str++;
+	*out = result;
+	return str;
+}
 
-		while (isdigit(*str)) {
-			if (len_fraction < numdig) {
-				fraction = fraction * base + (*str - '0');
-				dot_position++;
-			}
 
-			if (fraction) {
-				len_fraction++;
-			}
-
-			is_number = 1;
+static const char *scan_number_long_double(const char *str, bool hex, size_t digit_limit, size_t *n_decimal, size_t *n_overflowed, long double *out)
+{
+	const int base = hex ? 16 : 10;
+	long double result = 0;
+	bool got_dec = false;
+	size_t n_digits = 0;
+	while (1) {
+		char c = *str;
+		if (!got_dec && (c == DECIMAL_POINT)) {
+			got_dec = true;
 			str++;
+			continue;
 		}
-	}
 
-	/*  get the exponent part */
-	if (*str == 'E' || *str == 'e') {
-		str++;
+		int dig = get_digit_value(c, hex);
+		if (dig < 0) {
+			break;
+		}
 
-		if (*str == '-') {
-			exp_negative = 1;
-			str++;
+		if (n_digits <= digit_limit) {
+			n_digits++;
+			result = result * base + dig;
+			if (got_dec) {
+				(*n_decimal)++;
+			}
 		}
-		else if (*str == '+') {
-			str++;
+		else if (!got_dec) {
+			(*n_overflowed)++;
 		}
 
-		if (!isdigit(*str)) {
-			str--;
+		str++;
+	}
 
-			if (*str == '+' || *str == '-') {
-				str--;
+	*out = result;
+	return str;
+}
+
+static long double mul_exp10(long double in, int64_t exp)
```

Comment:
Do you assume the value of `FLT_ROUNDS`? Which rounding direction do you use? By default (in our toolchain) `FLT_ROUNDS` is `FE_TONEAREST`. If that's what you implement, then maybe you could add some `#warning` if `FLT_ROUNDS != FE_TONEAREST`. Of course, implementing other rounding modes is unnecessary, but this check will avoid confusion if we ever use another toolchain and `FLT_ROUNDS` changes.

Reply:
This code was tested with `FLT_ROUNDS == FE_TONEAREST`. I added a warning in case this is changed. So far setting the rounding mode is not implemented in libphoenix.